12.6.6.1: Defensive copy of input parameter

**Describe the bug**

In the C# 7.x draft, 12.6.6.1 (Function member invocation / General) describes how a copy of a value-type instance "not classified as a variable" is made and used instead of the original.  This defensive copying should also apply to an input parameter unless its type is a `readonly struct`.

**Example**

```csharp
using System;
struct S {
    public int I;
    public void M() { I++; }
}
readonly struct R {
    public void M() {}
}
class C {
    static void Main() {
        N(new S(), new S(), new R());
    }
    static void N(S s1, in S s2, in R r) {
        s1.M(); // no copying
        Console.WriteLine(s1.I); // 1
        
        s2.M(); // copy required
        Console.WriteLine(s2.I); // 0

        r.M(); // no copying
    }
}
```

**Expected behavior**

In 12.6.6.1, declare:

> If `E` is not classified as a variable, <ins>or if `E` is an input parameter (15.6.2.3) and `V` is not a readonly struct type (16.2.2),</ins> then a temporary local variable of E’s type is created and the value of E is assigned to that variable.

**Additional context**

None.


---
[Associated WorkItem - 157261](https://dev.azure.com/msft-skilling/Content/_workitems/edit/157261)

> * Member access `e.F` where F is a field and `e` is either a readonly _variable_reference_, or not a _variable_reference_ at all (e.g. a method call that returns by value).

Huh, apparently `e` still needs to be a *variable_reference*:

```csharp
partial struct S {
    public int I;
}
static class Ex {
    public static ref readonly S GetThat(this in S that) => ref that; // OK
}
partial class C {
    S M1() => new S { I = 5 };
    void M2(in int i) {}
    int M3() => M1().I;
    void M4() {
        //M2(in M1().I); // error CS1612
    }
    void M5() {
        M2(in M1().GetThat().I); // OK, copy M1() to local
    }
}

#if false // C# 11 Low Level Struct Improvements
partial struct S {
    [System.Diagnostics.CodeAnalysis.UnscopedRef]
    public ref readonly S GetThis() => ref this;
}
partial class C {
    void M6() {
        M2(in M1().GetThis().I); // OK, copy M1() to local
    }
}
#endif
```

--- a/standard/expressions.md
+++ b/standard/expressions.md
@@ -1211,7 +1211,11 @@ The run-time processing of a function member invocation consists of the followin
   - `M` is invoked.
 - Otherwise, if the type of `E` is a value-type `V`, and `M` is declared or overridden in `V`:
   - `E` is evaluated. If this evaluation causes an exception, then no further steps are executed. For an instance constructor, this evaluation consists of allocating storage (typically from an execution stack) for the new object. In this case `E` is classified as a variable.
-  - If `E` is not classified as a variable, or if `E` is an input parameter (§15.6.2.3) and `V` is not a readonly struct type (§16.2.2), then a temporary local variable of `E`’s type is created and the value of `E` is assigned to that variable. `E` is then reclassified as a reference to that temporary local variable. The temporary variable is accessible as `this` within `M`, but not in any other way. Thus, only when `E` is a true variable is it possible for the caller to observe the changes that `M` makes to `this`.
+  - If `E` is not classified as a variable, or if `V` is not a readonly struct type (§16.2.2), and `E` is one of:
+    - an input parameter (§15.6.2.3), or
+    - a `readonly` field (§15.5.3), or
+    - a `readonly` reference variable or return (§9.7),
+    then a temporary local variable of `E`’s type is created and the value of `E` is assigned to that variable. `E` is then reclassified as a reference to that temporary local variable. The temporary variable is accessible as `this` within `M`, but not in any other way. Thus, only when `E` is a true variable is it possible for the caller to observe the changes that `M` makes to `this`.
   - The argument list is evaluated as described in [§12.6.2](expressions.md#1262-argument-lists).
   - `M` is invoked. The variable referenced by `E` becomes the variable referenced by `this`.
 - Otherwise:

--- a/standard/expressions.md
+++ b/standard/expressions.md
@@ -1216,7 +1216,7 @@ The run-time processing of a function member invocation consists of the followin
     - a `readonly` field (§15.5.3), or
     - a `readonly` reference variable or return (§9.7),
 
-  then a temporary local variable of `E`’s type is created and the value of `E` is assigned to that variable. `E` is then reclassified as a reference to that temporary local variable. The temporary variable is accessible as `this` within `M`, but not in any other way. Thus, only when `E` is a true variable is it possible for the caller to observe the changes that `M` makes to `this`.
+  then a temporary local variable of `E`’s type is created and the value of `E` is assigned to that variable. `E` is then reclassified as a reference to that temporary local variable. The temporary variable is accessible as `this` within `M`, but not in any other way. Thus, only when `E` can be written is it possible for the caller to observe the changes that `M` makes to `this`.
   - The argument list is evaluated as described in [§12.6.2](expressions.md#1262-argument-lists).
   - `M` is invoked. The variable referenced by `E` becomes the variable referenced by `this`.
 - Otherwise:

--- a/standard/expressions.md
+++ b/standard/expressions.md
@@ -1211,7 +1211,7 @@ The run-time processing of a function member invocation consists of the followin
   - `M` is invoked.
 - Otherwise, if the type of `E` is a value-type `V`, and `M` is declared or overridden in `V`:
   - `E` is evaluated. If this evaluation causes an exception, then no further steps are executed. For an instance constructor, this evaluation consists of allocating storage (typically from an execution stack) for the new object. In this case `E` is classified as a variable.
-  - If `E` is not classified as a variable, then a temporary local variable of `E`’s type is created and the value of `E` is assigned to that variable. `E` is then reclassified as a reference to that temporary local variable. The temporary variable is accessible as `this` within `M`, but not in any other way. Thus, only when `E` is a true variable is it possible for the caller to observe the changes that `M` makes to `this`.
+  - If `E` is not classified as a variable, or if `E` is an input parameter (§15.6.2.3) and `V` is not a readonly struct type (§16.2.2), then a temporary local variable of `E`’s type is created and the value of `E` is assigned to that variable. `E` is then reclassified as a reference to that temporary local variable. The temporary variable is accessible as `this` within `M`, but not in any other way. Thus, only when `E` is a true variable is it possible for the caller to observe the changes that `M` makes to `this`.
   - The argument list is evaluated as described in [§12.6.2](expressions.md#1262-argument-lists).
   - `M` is invoked. The variable referenced by `E` becomes the variable referenced by `this`.
 - Otherwise:

--- a/standard/expressions.md
+++ b/standard/expressions.md
@@ -1215,7 +1215,8 @@ The run-time processing of a function member invocation consists of the followin
     - an input parameter (§15.6.2.3), or
     - a `readonly` field (§15.5.3), or
     - a `readonly` reference variable or return (§9.7),
-    then a temporary local variable of `E`’s type is created and the value of `E` is assigned to that variable. `E` is then reclassified as a reference to that temporary local variable. The temporary variable is accessible as `this` within `M`, but not in any other way. Thus, only when `E` is a true variable is it possible for the caller to observe the changes that `M` makes to `this`.
+
+  then a temporary local variable of `E`’s type is created and the value of `E` is assigned to that variable. `E` is then reclassified as a reference to that temporary local variable. The temporary variable is accessible as `this` within `M`, but not in any other way. Thus, only when `E` is a true variable is it possible for the caller to observe the changes that `M` makes to `this`.
   - The argument list is evaluated as described in [§12.6.2](expressions.md#1262-argument-lists).
   - `M` is invoked. The variable referenced by `E` becomes the variable referenced by `this`.
 - Otherwise:

